Use FileChannel.transferTo() to send static content

sbordet opened the pull request.

sbordet committed:
Experimental sketch for #13392.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
More experiments for #13392.

Implemented `Response.write(Response, boolean, Content.Source, Callback)`.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
More experiments for #13392.

Implemented `Sink.write(Response, boolean, Content.Source, Callback)`.
Now the API is just the `Sink.write()` above, and most of the rest is
hidden.

Updated IOResources to use the new method above, so normal file serving
should be able to leverage zero-copy.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Added TODOs.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
More work in progress.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Started analysis for HTTP/2.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

gregw commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/Content.java:
```diff
@@ -175,6 +208,13 @@ interface Factory
+            Source getContentSource();
+
+            void setContentSource(Source source);
+        }
```
See #13656 as an alternate way to tunnel a Source through a ByteBuffer only API.
Although I'm hating the current iteration of Aware less than I did originally

gregw replied:
I don't like my tunnel approach now

gregw commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/Content.java:
```diff
@@ -866,6 +908,36 @@ static void write(Sink sink, boolean last, String utf8Content, Callback callback
+            {
+                if (sink instanceof Content.Source.Aware aware)
+                    return aware;
+                if (sink instanceof Wrapper wrapper)
```
I think if the sink is wrapped then we should not bypass it with transferTo
Perhaps we need an AwareWrapper, that would allow itself to be bypassed?

gregw commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/Content.java:
```diff
@@ -866,6 +908,36 @@ static void write(Sink sink, boolean last, String utf8Content, Callback callback
+            {
+                // Optimization to enable zero-copy.
+                aware.setContentSource(source);
+                sink.write(last, TRANSFER, callback);
```
I feel that we should allow some failures on the callback here to fall through to a normal copy.  I.e. if something in the implementation is Aware but cannot do a transferTo (perhaps because it would be non optimal) then it can fail this write with a `TransferToFailed` exception, which would then just do the normal copy below.

gregw commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/Content.java:
```diff
@@ -866,6 +908,36 @@ static void write(Sink sink, boolean last, String utf8Content, Callback callback
+            Content.Source.Aware aware = findContentSourceAware(sink);
+            if (aware != null)
+            {
+                // Optimization to enable zero-copy.
```
Explain in the comment how the optimization still uses the sink.write path, so that commit logic etc. can be triggered.    Specifically, we do not do `aware.transfer(source, callback)` because we might need to commit the response.

gregw commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java:
```diff
@@ -1320,17 +1323,19 @@ else if (last && !(totalWritten == 0 && HttpMethod.HEAD.is(_request.getMethod())
                     return;
                 }
 
+                // TODO: check that we don't have both a ByteBuffer and a Content.Source.
```
I think you should also check `content==TRANSFER`, because if some sink wrapper inserted different content, then the presence of the source will need to be ignored

gregw reviewed (changes requested):
I'm hating this less

sbordet committed:
Implemented transferTo() for HTTP/2.

TODO:
* Review ByteChannelContentSource and subclasses for test failures.
* Implement SeekableByteChannelContentSource.rewind().
* Review Handler.Wrappers that wrap Response to be aware of Sink.TRANSFER_TO.
* Write tests to cover all the branches introduced, for example:
** Content.transfer()
** Content.write(Sink, boolean, Source, Callback)
** Write some ByteBuffer, then a transferable source, and viceversa.
** Etc.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>
